agent/worker.go

```diff
@@ -409,6 +419,49 @@ func reconcileConfigs(ctx context.Context, w *worker, assignments []*api.Assignm
 	return nil
 }
 
+func reconcileVolumes(ctx context.Context, w *worker, assignments []*api.AssignmentChange, fullSnapshot bool) error {
+	var (
+		updatedVolumes []api.VolumeAssignment
+		removedVolumes []string
+	)
+	for _, a := range assignments {
+		if r := a.Assignment.GetVolume(); r != nil {
+			switch a.Action {
+			case api.AssignmentChange_AssignmentActionUpdate:
+				updatedVolumes = append(updatedVolumes, *r)
+			case api.AssignmentChange_AssignmentActionRemove:
+				removedVolumes = append(removedVolumes, r.ID)
+			}
+
+		}
+	}
+
+	volumesProvider, ok := w.executor.(exec.VolumesProvider)
+	if !ok {
+		if len(updatedVolumes) != 0 || len(removedVolumes) != 0 {
+			log.G(ctx).Warn("volumes update ignored; executor does not support volumes")
+		}
+		return nil
+	}
+
+	volumes := volumesProvider.Volumes()
+
+	log.G(ctx).WithFields(logrus.Fields{
+		"len(updatedVolumes)": len(updatedVolumes),
+		"len(removedVolumes)": len(removedVolumes),
+	}).Debug("(*worker).reconcileVolumes")
+
+	// If this was a complete set of volumes, we're going to clear the volumes map and add all of them
+	if fullSnapshot {
+		volumes.Reset()
```

Comment:
Add another TODO comment here, because when we reset volumes, we don't want to unpublish and unstage only to publish and stage again a moment later. That works for secrets and configs because the object itself contains everything you need, but for volumes it would be a mess.